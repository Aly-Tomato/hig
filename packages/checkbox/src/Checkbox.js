import React, { Component } from "react";
import PropTypes from "prop-types";

import CheckboxPresenter from "./CheckboxPresenter/CheckboxPresenter";

export default class Checkbox extends Component {
  static propTypes = {
    /**
     * Checks the checkbox
     */
    checked: PropTypes.bool,
    /**
     * Initially checks the checkbox, but allows user action to change it
     */
    defaultChecked: PropTypes.bool,
    /**
     * Prevents user actions on the checkbox
     */
    disabled: PropTypes.bool,
    /**
     * Sets indeterminate state for checkbox
     */
    indeterminate: PropTypes.bool,
    /**
     * Text identifying the field
     */
    label: PropTypes.string,
    /**
     * The name of the checkbox as submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called when user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user puts focus on the field
     */
    onFocus: PropTypes.func,
    /**
     * Marks the field as required, text shown to explain requirement
     */
    required: PropTypes.string,
    /**
     * Value submitted with a form if checked
     */
    value: PropTypes.string
  };

  static defaultProps = {
    defaultChecked: true
  };

  state = {
    checked: this.props.defaultChecked
  };

  /**
   * @param {MouseEvent} event
   */
  handleClick = event => {
    if (this.props.onChange) {
      this.props.onChange(event);
    }

    this.setState({
      checked: !this.state.checked
    });
  };

  renderChecked() {
    if (this.props.checked !== undefined) {
      return this.props.checked;
    }
    return this.state.checked;
  }

  render() {
    return (
      <CheckboxPresenter
        {...this.props}
        checked={this.renderChecked()}
        onChange={this.handleClick}
      />
    );
  }
}
