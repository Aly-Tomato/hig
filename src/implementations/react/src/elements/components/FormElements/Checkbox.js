import React from "react";
import CheckboxAdapter from "../../../adapters/FormElements/CheckboxAdapter";
import * as PropTypes from "prop-types";

class Checkbox extends React.Component {
  constructor(props) {
    super(props);

    const controlled = props.checked === undefined ? false : true;

    this.state = {
      checked: this.getDefaultChecked(),
      controlled
    };
  }

  static defaultProps = {
    onChange: () => {}
  };

  getDefaultChecked() {
    const { defaultChecked, checked } = this.props;

    if (checked !== undefined) {
      return checked;
    } else if (defaultChecked !== undefined) {
      return defaultChecked;
    } else {
      return false;
    }
  }

  getRenderedChecked() {
    const { checked } = this.props;

    if (checked !== undefined) {
      return checked;
    } else {
      return this.state.checked;
    }
  }

  handleChange = event => {
    this.props.onChange(event);
    if (this.checkboxEl) {
      if (this.state.controlled) {
        this.checkboxEl.forceNextReset();
        this.setState({ checked: this.state.checked });
      } else  {
        this.setState({ checked: event.target.checked });
      }
    }
  };

  setCheckboxEl = checkboxEl => {
    this.checkboxEl = checkboxEl;
  };

  render() {
    return (
      <CheckboxAdapter
        ref={this.setCheckboxEl}
        checked={this.getRenderedChecked()}
        disabled={this.props.disabled}
        name={this.props.name}
        label={this.props.label}
        required={this.props.required}
        value={this.props.value}
        onChange={this.handleChange}
        onFocus={this.props.onFocus}
        onHover={this.props.onHover}
      />
    );
  }
}

Checkbox.__docgenInfo = {
  props: {
    checked: {
      description: "boolean - sets whether the checkbox is checked"
    },
    disabled: {
      description: "boolean - sets whether the checkbox is disabled"
    },
    label: {
      description: "sets the label text for the checkbox"
    },
    name: {
      description: "sets the name attribute of the checkbox input"
    },
    required: {
      description: "string - sets the whether the checkbox is required and displays the provided message"
    },
    value: {
      description: "sets the value attribute of the checkbox input"
    },
    onHover: {
      description: "triggers when you hover over the button"
    },
    onChange: {
      description: "triggers when you check/uncheck the checkbox"
    },
    onFocus: {
      description: "triggers the checkbox component receives focus"
    }
  }
};

Checkbox.propTypes = {
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  label: PropTypes.string,
  name: PropTypes.string,
  required: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onHover: PropTypes.func
};

export default Checkbox;
