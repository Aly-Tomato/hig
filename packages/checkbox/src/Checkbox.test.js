import React from "react";
import { mount, render } from "enzyme";
import Checkbox from "./Checkbox";

describe("Checkbox", () => {
  const subject = (props = {}) => mount(<Checkbox title="Checkbox" {...props} />);
  let wrapper;
  let checkbox;

  describe("integration", () => {
    beforeEach(() => {
      wrapper = render(<Checkbox checked={true}/>);
      checkbox = wrapper.find(".hig__input-button");
    });

    it("renders a presenter", () => {
      expect(checkbox.length).not.toBe(0);
    });

    it("input element has checked class when props.checked is true", () => {
      expect(checkbox.hasClass("hig__input-button--checked"));
    });
  });

  describe("event handlers", () => {
    let eventHandler;

    describe("onBlur", () => {
      beforeEach(() => {
        eventHandler = jest.fn();
        wrapper = mount(<Checkbox onBlur={eventHandler} />);
      });

      it("is triggered on blur", () => {
        wrapper.find("input").simulate("blur");
        expect(eventHandler).toBeCalled();
      });
    });

    describe("onChange", () => {
      beforeEach(() => {
        eventHandler = jest.fn();
        wrapper = mount(<Checkbox onChange={eventHandler} />);
      });

      it("is triggered on change", () => {
        wrapper.find("input").simulate("change");

        expect(eventHandler).toHaveBeenCalled();
      });
    });

    describe("onFocus", () => {
      beforeEach(() => {
        eventHandler = jest.fn();
        wrapper = subject({ onFocus: eventHandler });
      });

      it("is triggered on focus", () => {
        wrapper.find("input").simulate("focus");
        expect(eventHandler).toBeCalled();
      });
    });
  });
});
