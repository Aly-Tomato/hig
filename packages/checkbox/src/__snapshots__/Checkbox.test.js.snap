// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox/Checkbox integration renders correctly 1`] = `
.emotion-6 {
  box-sizing: border-box;
  display: inline-block;
  position: relative;
  height: 16px;
  width: 16px;
}

.emotion-0 {
  opacity: 0;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  z-index: 1;
  padding: 0;
  margin: 0;
}

.emotion-5 {
  position: relative;
  display: block;
  box-sizing: border-box;
  color: transparent;
  border: 1px solid rgba(190,200,210,0.5);
  background-color: #ffffff;
  text-align: center;
  font-size: 25px;
  line-height: 14px;
  height: 16px;
  width: 16px;
  border-radius: 3px;
  cursor: default;
}

.emotion-5 svg {
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%);
  -ms-transform: translateX(-50%) translateY(-50%);
  transform: translateX(-50%) translateY(-50%);
}

.emotion-5 svg * {
  fill: #ffffff;
}

.emotion-2 {
  display: none;
}

.emotion-1 {
  fill: #808080;
}

.emotion-1 > * {
  fill: #808080;
}

<div
  className="emotion-6"
>
  <input
    checked={false}
    className="emotion-0"
    disabled={false}
    label="HIG Checkbox"
    name="checkbox"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={undefined}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    type="checkbox"
    value="value"
  />
  <span
    className="emotion-5"
  >
    <span
      className="emotion-2"
    >
      <svg
        className="emotion-1"
        height="16px"
        viewBox="0 0 16 16"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.2 12L2 8.1l1.3-1.3 2.9 2.7L12.7 3 14 4.3z"
        />
      </svg>
    </span>
    <span
      className="emotion-2"
    >
      <svg
        className="emotion-1"
        height="2px"
        viewBox="0 0 8 2"
        width="8px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Rectangle 11
        </title>
        <path
          d="M0 0h8v2H0z"
          fill="#FFF"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </span>
</div>
`;
